src/ZoneServer/World/Spawner/MonsterSpawner.cs

```diff
@@ -110,35 +110,30 @@ public class MonsterSpawner
 		/// <summary>
 		/// Creates new instance.
 		/// </summary>
+		/// <param name="name"></param>
 		/// <param name="monsterClassId"></param>
 		/// <param name="minAmount"></param>
 		/// <param name="maxAmount"></param>
-		/// <param name="mapClassName"></param>
-		/// <param name="area"></param>
 		/// <param name="initialSpawnDelay"></param>
 		/// <param name="minRespawnDelay"></param>
 		/// <param name="maxRespawnDelay"></param>
 		/// <param name="tendency"></param>
 		/// <param name="propertyOverrides"></param>
-		public MonsterSpawner(int monsterClassId, int minAmount, int maxAmount, string mapClassName, IShape area, TimeSpan initialSpawnDelay, TimeSpan minRespawnDelay, TimeSpan maxRespawnDelay, TendencyType tendency, PropertyOverrides propertyOverrides)
+		public MonsterSpawner(string name, int monsterClassId, int minAmount, int maxAmount, TimeSpan initialSpawnDelay, TimeSpan minRespawnDelay, TimeSpan maxRespawnDelay, TendencyType tendency, PropertyOverrides propertyOverrides)
```

Comment:
Hm... So you're assigning names to the spawners and then assign spawn points to them. That is certainly one way of doing it, but wouldn't it be more flexible to treat the spawn points as a separate thing, a list of areas that can be used for arbitrary purposes? For example, let's assume that the spawners and ambushes use the same areas. With the current approach, we would have to assign the same areas again to the ambushes, but if the spawn points are just a list, stored somewhere in the server, we could access them for ambushes as well, and whatever other things we might come up with. What do you think?

Reply:
I thought a lot about this but I don't see how to implement your suggestion in a neat, organized way.

Let's assume, for example, that the spawn point list is stored as a property of Map class, because that would make the most sense. Then we arrive at the following issues:

- The spawner class is not tied to a map, so it won't know which map to look for when iterating over spawn points.
- The spawner class would then have to have a list of spawn points in its property, making it redundant to have list of spawn points also in map class property (not that I'm against this redundancy, but I think the list in map will just not be used).
- Spawn points are identified by their map and area. There isn't a name string to associate them to (like the spawner class). So re-using them on scripts as you intend is quite complicated. I also wouldn't want to add name strings to spawn points, it would be too unecessarily confusing. Perhaps on spawn point collections, but that's not under the scope of this PR.
- I imagine the spawn point class as a "component" of spawner or (the future) ambush spawner classes. It should not exist on it's own, as it makes no sense to have a spawn point in a map that does nothing.
- For future classes that would use same spawn point (i.e. ambush spawner class) we could just have a shortcut that adds the spawn point to the 'spawner list' and 'ambush spawner list'. No real need to add spawn point to the world or map and then an additional effort to identify and retrieve it to be re-used.

In short, I believe having each spawner class (spawner & ambush spawner) having their own internal list of spawn points as a better architecture. But what do you think?

Reply:
I agree that each spawner should have its own list, particularly because I also imagine that it could be useful to be able to customize a pre-defined list. Though I imagine it wouldn't be very complicated. The way I picture it, the spawn points would be placed on World because they don't necessarily have a relation to any one map. It'd basically be like a database that we generate at run-time and then reference to build a spawner's list of potential spawn areas.
```cs
World:
    SpawnAreaCollection SpawnAreas
    SpawnerCollection Spawners

Shortcuts:
    AddSpawnArea(string spawnerAreaListName, string mapClassName, ...)
        World.SpawnAreas.GetOrCreateList(spawnerAreaListName).Add(...)

    SpawnArea[] SpawnAreas(string spawnAreaListName)
        return World.SpawnAreas.Get(spawnAreaListName)

    AddSpawner(int monsterClassId, ... string mapClassName, IShapeF area, ...))
        AddSpawner(monsterClassId, ... new[] { new SpawnArea(mapClassName, area) }, ...)

    AddSpawner(int monsterClassId, ... SpawnArea[] spawnAreas, ...))
        World.Spawners.Add(...)

// Usage Examples
AddSpawnArea("FSiauliaiWestOnions", "f_siauliai_west", ...)
AddSpawnArea("FSiauliaiWestOnions", "f_siauliai_west", ...)

AddSpawner(MonsterId.Onion, ... "f_siauliai_west", Area(...), ...)
AddSpawner(MonsterId.Onion, ... SpawnAreas("FSiauliaiWestOnions"), ...)
```
An alternative to using named lists would be to create local lists that can be directly passed to `AddSpawner`.
```cs
var onionAreas = CreateSpawnerAreas()
onionAreas.Add("f_siauliai_west", ...)

AddSpawner(MonsterId.Onion, ... onionAreas, ...)
```
The disadvantage is that we wouldn't be able to easily reuse spawn area lists from other scripts, but I guess that would be rare. Well, unless you wanted to add a custom spawner that reuses an existing list 🤔 For example, I might want to add a regular spawn for a Large Kepa, as a kind of mini boss, and have it spawn on one of the Kepa spawns in West Siauliai. If it's not possible to reference the existing list, I'd have to duplicate it. Granted, that's not an uncommon thing to do in such a case, but the list _could_ be made accessible.

> Spawn points are identified by their map and area

Each independent spawn point doesn't need an identifier I don't think, but yea, the _list_ of spawn points would require one if we wanted to be able to query it as needed for whatever purposes, be it as a reference for the spawn areas or for whatever other purposes someone might come up with.

> There isn't a name string to associate them to (like the spawner class)

Right, but we don't actually _need_ to name the spawners, do we? I can't think of a general reason to do that off the top of my head. It's just the implementation that made sense if you want to assign spawn points to the spawners instead of the other way around. Not saying this doesn't work, but you could just as well name spawn area lists and remove the spawner names.

> as it makes no sense to have a spawn point in a map that does nothing

I hope my comparison to a database makes sense here. It's somewhat like the resurrection point database, just a list of named lists of map/area values. We could technically also actually put that data into a database, like the game does, though I think we want to create it at run-time for customization purposes.

> For future classes that would use same spawn point (i.e. ambush spawner class) we could just have a shortcut that adds the spawn point to the 'spawner list' and 'ambush spawner list'. 

Would you then give the spawners and the ambushes the same name? You'd have to assign the same lists to the same spawners and ambushes after all. Doesn't that seem confusing? And another potential problem I just remembered mentioning before, what if you want to use the same spawn areas for multiple monsters?
```cs
AddSpawner("FSiauliaiWestOnions", MonsterId.Onion, ...)
AddSpawner("FSiauliaiWestOnions", MonsterId.Leaf_diving, ...)

AddAmbush("FSiauliaiWestOnions", MonsterId.Onion, ...)

AddSpawnPoint("FSiauliaiWestOnions", "f_siauliai_west", ...)
AddSpawnPoint("FSiauliaiWestOnions", "f_siauliai_west", ...)
AddSpawnPoint("FSiauliaiWestOnions", "f_siauliai_west", ...)
AddSpawnPoint("FSiauliaiWestOnions", "f_siauliai_west", ...)
```
I mean, that _will_ work if you iterate over all possible options inside `AddSpawnPoint`, but doesn't it feel a little weird for names not to be unique?

Reply:
I see what you're saying. I will add a 'MonsterSpawnPointCollection' class that has a list of 'MonsterSpawnPoint' in it and an identifier string. It makes more sense to have the identifier stirng in the spawn point collection than on the spawner. That way, you can associate multiple spawners to a spawn point collection and spawn different mob classes or add ambushes.

'MonsterSpawnPointCollection' will also be stored in world, as you recommended, because it can contain spawn points on different maps.

I was trying to get away with this PR without implementing spawn point collections, but I think you're right, that is the best architecture.

Reply:
I have added the functionality you suggested. Now you can assign multiple spawners to a same spawn point collection:
![image](https://github.com/NoCode-NoLife/melia/assets/91282765/11075b56-222f-47ea-a0fb-84637a2f5d9d)

And you can also do it even if you add the spawn point after:
![image](https://github.com/NoCode-NoLife/melia/assets/91282765/7b0f4e0a-da74-41a4-bdd2-c409bce7e0ad)

Essentially, the spawn point collection has a string name identifier and the spawner class receives that string and tries to resolve the object on runtime. The name identifier is stored at world so you can even use it cross-script. I also removed name from spawner class, as you recommended (now only the spawn point collection has a name, but the users making scripts don't need to know how it works internally).

Reply:
Maybe I should rename the "spawnPointCollectionName" variables to something more intuitive for users that produce scripts? I mean, for them, they don't even need to know spawn point collections exist. Maybe I should name it "spawnIdentifier" or something generic and abstract? What do you think?

Reply:
Sounds good!

Reply:
One thing I'm not sure about is the order of the `AddSpawnPoint` arguments. Don't you think putting the name at the end makes taking in the information more difficult? Isn't the name of the list the most important information in a call that adds a new element to said list? Say you're staring at a block of spawn point definitions, it becomes difficult to determine whether they even all go into the same list.
```cs
AddSpawnPoint("f_siauliai_west", Area(-555, -1105, -655, -1105, -655, -1205, -555, -1205), "SpawnPointCollection10.f_siaulia_west");
AddSpawnPoint("f_siauliai_west", Spot(-575, -1300), "SpawnPointCollection10.f_siaulia_west");
AddSpawnPoint("f_siauliai_west", Area(-555, -1105, -655, -1105, -655, -1205, -555, -1205, -555, -1205), "SpawnPointCollection11.f_siaulia_west");
AddSpawnPoint("f_siauliai_west", Spot(-575, -1300), "SpawnPointCollection10.f_siaulia_west");
AddSpawnPoint("f_siauliai_west", Rect(-555, -1105, 500), "SpawnPointCollection10.f_siaulia_west");
AddSpawnPoint("f_siauliai_west", Area(-555, -1105, -655, -1105, -655, -1205, -555, -1205), "SpawnPointCollection10.f_siaulia_west");
AddSpawnPoint("f_siauliai_west", Area(-555, -1105, -655, -1105, -655, -1205, -555, -1205, -555, -1205), "SpawnPointCollection11.f_siaulia_west");
AddSpawnPoint("f_siauliai_west", Spot(-575, -1300), "SpawnPointCollection10.f_siaulia_west");
```
The only information you can glean at a glance is that all of those are in West Siauliai, but not which lists they belong to. On the other hand, putting the name as the first argument lets you parse both the lists they belong to and the maps at once.
```cs
AddSpawnPoint("SpawnPointCollection10.f_siaulia_west", "f_siauliai_west", Area(-555, -1105, -655, -1105, -655, -1205, -555, -1205));
AddSpawnPoint("SpawnPointCollection10.f_siaulia_west", "f_siauliai_west", Spot(-575, -1300));
AddSpawnPoint("SpawnPointCollection11.f_siaulia_west", "f_siauliai_west", Area(-555, -1105, -655, -1105, -655, -1205, -555, -1205, -555, -1205));
AddSpawnPoint("SpawnPointCollection10.f_siaulia_west", "f_siauliai_west", Spot(-575, -1300));
AddSpawnPoint("SpawnPointCollection10.f_siaulia_west", "f_siauliai_west", Rect(-555, -1105, 500));
AddSpawnPoint("SpawnPointCollection10.f_siaulia_west", "f_siauliai_west", Area(-555, -1105, -655, -1105, -655, -1205, -555, -1205));
AddSpawnPoint("SpawnPointCollection11.f_siaulia_west", "f_siauliai_west", Area(-555, -1105, -655, -1105, -655, -1205, -555, -1205, -555, -1205));
AddSpawnPoint("SpawnPointCollection10.f_siaulia_west", "f_siauliai_west", Spot(-575, -1300));
```

Reply:
I thought of that too, and already changed it, just reviewing some spawn points now because some NPCs are being spawned because client data has them with "Monster" type ^^

![image](https://github.com/NoCode-NoLife/melia/assets/91282765/47e3319a-fa84-4551-bcfd-97225ad3d96a)

Reply:
I've finished my updates and removed some spawn points that were generating hidden triggers and npcs. Let me know if there's any other changes to be done in this PR.